shell/platform/android/io/flutter/embedding/FlutterFragment.java

```diff
@@ -0,0 +1,423 @@
+   *
+   * TODO(mattcarroll): speed aside, should developers be given the opportunity to use different Drawables?
+   *
+   * Returns null if no {@code windowBackground} is set for the activity.
+   */
+  @SuppressWarnings("deprecation")
+  @Nullable
+  private Drawable getSplashScreenDrawableFromActivityTheme() {
+    TypedValue typedValue = new TypedValue();
+    if (!getContextCompat().getTheme().resolveAttribute(
+        android.R.attr.windowBackground,
+        typedValue,
+        true)) {
+      return null;
+    }
+    if (typedValue.resourceId == 0) {
+      return null;
+    }
+    try {
+      return getResources().getDrawable(typedValue.resourceId);
+    } catch (Resources.NotFoundException e) {
+      Log.e(TAG, "Referenced launch screen windowBackground resource does not exist");
+      return null;
+    }
+  }
+
+  /**
+   * Returns the Flutter view used by this {@code Fragment}; will be null before
+   * {@link #onCreateView(LayoutInflater, ViewGroup, Bundle)} is invoked. Will be
+   * non-null after {@link #onCreateView(LayoutInflater, ViewGroup, Bundle)} is invoked, up
+   * until {@link #onDestroyView()} is invoked.
+   */
+  @SuppressWarnings("unused")
+  @Nullable
+  public FlutterView getFlutterView() {
+    return flutterView;
+  }
+
+  /**
+   * Starts running Dart within the FlutterView for the first time.
+   *
+   * The difference between the first run and subsequent reloads is that when FlutterView is
+   * run for the first time, it attempts to reuse an existing Dart isolate.
+   */
+  private void doInitialFlutterViewRun() {
+    runFlutterView(getInitialRoute(), getAppBundlePath(), true);
+  }
+
+  /**
+   * Reloads Dart code with that at the given {@code appBundlePath}, sets Flutter's initial route
+   * to {@code initialRoute}, and then reloads the Dart VM.
+   *
+   * TODO(mattcarroll): is the above description accurate?
+   *
+   * @param initialRoute initial Flutter route to display after reloading Dart code
+   * @param appBundlePath Android file path to Flutter's app bundle
+   */
+  public void hotReload(@Nullable String initialRoute, @NonNull String appBundlePath) {
+    runFlutterView(initialRoute, appBundlePath, false);
+  }
+
+  /**
+   * Instructs the FlutterView to start running with the given {@code appBundlePath}, and starting
+   * Flutter at the given {@code initialRoute}.
+   *
+   * @param initialRoute initial Flutter route to display after reloading Dart code
+   * @param appBundlePath Android file path to Flutter's app bundle
+   * @param reuseIsolate whether or not to reuse an existing Dart isolate
+   */
+  private void runFlutterView(@Nullable String initialRoute, @NonNull String appBundlePath, boolean reuseIsolate) {
+    if (initialRoute != null) {
+      flutterView.setInitialRoute(initialRoute);
+    }
+    // TODO: what does it really mean for this condition to be false? is that ok? should we log?
+    if (!flutterView.getFlutterNativeView().isApplicationRunning()) {
```

Comment:
This is intended to support apps that create a single Flutter engine instance and keep it running for the lifetime of the application.

In this mode, the engine is detached from the activity when the activity is destroyed, but the engine continues to execute Dart code.  When a new instance of the activity is created, the same engine is attached to the new activity.

See https://github.com/flutter/engine/commit/c7ab033f9e608fb55144739eb794e84a7281b5a7

Reply:
@jason-simmons Am I correct in thinking that we also shouldn't set the initial route on FlutterView if FlutterNativeView is already running?  Or should we go ahead and set the initial route in case it's hot restarted later?

Reply:
Also, another question, if we want to hot reload (like when we process an ACTION_RUN Intent), should we invoke runFromBundle even if the application is already running?

Reply:
I'm not sure what the user's goal would be if they used `setInitialRoute` in this scenario.

If the app has a custom `FlutterActivityDelegate.ViewFactory` that starts the Flutter engine before starting the activity, then the Dart code running in the engine has likely read the `ui.window.defaultRouteName` property before `setInitialRoute` can write to it.

I'm not familiar with the details of the `Naviagtor` on the framework side, but I doubt that it would make sense to overwrite `defaultRouteName` after the app is already running.  So I guess the activity can skip the `setInitialRoute` operation in this case.

Reply:
What about in the context of a hot restart?  If the initial route were changed on the Android side and then a hot restart was executed, would the restart then begin at the new initial route?  Or, more importantly, *should* it?

Reply:
@Hixie what are the use cases for `flutter run --route`?  Is it mostly intended for use in tests?

Reply:
It's mostly intended for development. It's a useful way to jump to a particular part of the app on startup. It dates from before we had hot reload. It's still useful in scenarios where hot reload doesn't work, though. It exercises similar paths to "deep linking".